cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,252 @@
+    denominator = tuple(a.args[0] for a in expr.args if isinstance(a, Pow) and a.args[1] == -1.0)
+
+    # Not a devision or does not have exp
+    if len(denominator) == 0 or len(numerator) == 0 or not expr.has(exp_function):
+        return None, None, None
+
+    (Vmin, Vmax, sp) = None, None, None
+    # U might be on top, try numerator / denominator and denominator / numerator
+    for num, denom in ((numerator, denominator), (denominator, numerator)):
+        found_on_top = False
+        for d in denom:  # Check arguments in denominator (or numerator)
+            if d.has(exp_function):
+                find_U = d.match(exp_function(U_wildcard) * -Z + 1.0)  # look for exp(U) * -Z + 1.0
+                if not check_bottom_match(find_U):
+                    find_U = d.match(exp_function(U_wildcard) * Z - 1.0)  # look for exp(U) * Z - 1.0
+                if check_bottom_match(find_U):
+                    # We found a match, since exp(U) * Z == exp(U + log(Z)) we can bring Z into the u expression
+                    u = (find_U[U_wildcard] + log(find_U[Z]))
+                    try:
+                        # Find the singularity point by solving u for V==0, excluding irratinal results
+                        sp = tuple(filter(lambda s: not s.has(I), solveset(u, V)))
+                        if sp:
+                            # we found a singularity point, now find Vmin, Vmax
+                            # by solving for U_offset either side of U
+                            find_v_low = solveset(u + U_offset, V)
+                            find_v_up = solveset(u - U_offset, V)
+                            find_v_low = tuple(find_v_low)
+                            find_v_up = tuple(find_v_up)
+                            assert find_v_low and find_v_up and len(find_v_low) == len(find_v_up) == len(sp) == 1, \
+                                'Expecting exactly 1 singularity point '
+                            (Vmin, Vmax, sp) = (find_v_low[0], find_v_up[0], sp[0])
+                    except TypeError:
+                        pass  # Result could be 'ConditionSet' which is not iterable and not Real
+
+                if Vmin is not None:  # check top
+                    for n in num:  # Check arguments in numerator (or denominator)
+                        match = n.match(Z * V - Z * SP_wildcard)  # search for a multiple of V - sp
+                        found_on_top = check_top_match(match, sp)
+                        if not found_on_top:
+                            # search for a exp(multiple of V - sp)
+                            match = n.match(exp_function(Z * V - Z * SP_wildcard))
+                            found_on_top = check_top_match(match, sp)
+                        if found_on_top:  # We've found a match stop looking in the other numerator arguments
+                            break
+                    if found_on_top:  # found singularity, no need to try further
```

Comment:
What if there's two singularities in an equation?

Reply:
we can deal with `singularity + singularity + ...` and find both since both parts of the addition are treated seperately. 

But to be fair it is true we don't really deal with `singularity * singularity `. 

I ran some tests as follows (with num_singularities initialised to 0 before looping starts):

It counts and removes the simularity found from denom/numerator to prevent finding it when we swap numerator & denomenator.
```
                if found_on_top:  # found singularity, no need to try further
                    num_singularities +=1
                    assert num_singularities < 2
                    if denom == denominator:
                        denominator = [v for v in denominator if v != d]
                    else:
                        numerator = [v for v in numerator if v != d]
```

This is of course an ugly check. Let me know what you think.

Reply:
Ok so I have now (yet to be pushed) made a fix and added a test, also fixed for the possibility of U==0 having multiple solutions